appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/listener/LocalTxConnectionEventListener.java

```diff
@@ -36,100 +36,138 @@
  * and therefore, elected the GPL Version 2 license, then the option applies
  * only if the new code is made subject to such option by the copyright
  * holder.
+ * 
+ * Portions Copyright [2018] [Payara Foundation and/or its affiliates]
  */
 
 package com.sun.enterprise.resource.listener;
 
-import com.sun.enterprise.resource.pool.PoolManager;
-import com.sun.enterprise.connectors.ConnectorRuntime;
-import com.sun.enterprise.resource.*;
+import static java.util.logging.Level.FINE;
 
-import javax.resource.spi.*;
-import java.util.*;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.logging.Logger;
+
+import javax.resource.ResourceException;
+import javax.resource.spi.ConnectionEvent;
+import javax.resource.spi.ManagedConnection;
+
+import com.sun.enterprise.connectors.ConnectorRuntime;
+import com.sun.enterprise.resource.ResourceHandle;
+import com.sun.enterprise.resource.pool.PoolManager;
+import com.sun.logging.LogDomains;
 
 /**
  * @author Binod P.G
  */
 public class LocalTxConnectionEventListener extends ConnectionEventListener {
 
-    private PoolManager poolMgr;
+    private static final Logger LOGGER = LogDomains.getLogger(LocalTxConnectionEventListener.class,
+            LogDomains.RSR_LOGGER);
+
+    private final PoolManager poolMgr;
 
     // connectionHandle -> ResourceHandle
     // Whenever a connection is associated with a ManagedConnection,
     // that connection and the resourcehandle associated with its
     // original ManagedConnection will be put in this table.
-    private IdentityHashMap associatedHandles;
+    private final Map<Object, ResourceHandle> associatedHandles;
+
+    private final ResourceHandle resource;
 
-    private ResourceHandle resource;
-        
     public LocalTxConnectionEventListener(ResourceHandle resource) {
         this.resource = resource;
-        this.associatedHandles = new IdentityHashMap(10);
+        this.associatedHandles = new IdentityHashMap<>(10);
```

Comment:
Would it be a sufficient enough synchronization to use Collections.synchronizedMap(new IdentityMap(10)) instead of synchronizing the whole listener?

Reply:
The reason I've not used a synchronized map and instead synchronized specific blocks is that I believe the poolManager.resourceClosed method needs to be run in the same atomic action, and a synchronized map wouldn't do this.

Reply:
Looks good then.